Async sendable

Geobert opened the pull request:
PR for discussion

Geobert commented:
Removing `?Send` leads me to add `+ Send`  to `dyn Request`, which leads me to add `+Send` again to `dyn WebRequest` which lead me to be stuck on:
```
error[E0277]: `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)` cannot be shared between threads safely
   --> oxide-auth-async\src\endpoint\access_token.rs:126:13
    |
126 |             &WrappedRequest::new(&mut request, self.allow_credentials_in_body),
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)` cannot be shared between threads safely
    |
    = help: the trait `std::marker::Sync` is not implemented for `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)`
    = note: required because of the requirements on the impl of `std::marker::Send` for `&(dyn oxide_auth::endpoint::query::QueryParameter + 'static)`
    = note: required because it appears within the type `std::borrow::Cow<'_, (dyn oxide_auth::endpoint::query::QueryParameter + 'static)>`
    = note: required because it appears within the type `endpoint::access_token::WrappedRequest<'_, R>`
    = note: required for the cast to the object type `dyn oxide_auth::code_grant::accesstoken::Request + std::marker::Send`
```

Because of:
```rust
// oxide_auth_async::endpoint::access_token
struct WrappedRequest<'a, R: WebRequest + Send + 'a> {
    /// Original request.
    request: PhantomData<R>,

    /// The query in the url.
    body: Cow<'a, dyn QueryParameter + 'static>,

    /// The authorization tuple
    authorization: Option<Authorization>,

    /// An error if one occurred.
    error: Option<FailParse<R::Error>>,

    /// The credentials-in-body flag from the flow.
    allow_credentials_in_body: bool,
}
```

But if I add ~`+Send`~ `+Sync` (as it's what's missing) to `dyn QueryParameter`:
```
error[E0277]: the trait bound `(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync+ 'static): std::clone::Clone` is not satisfied
  --> oxide-auth-async\src\endpoint\access_token.rs:47:5
   |
47 |     body: Cow<'a, dyn QueryParameter + Sync + 'static>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for 
`(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Send + 'static)`
   |
   = note: required because of the requirements on the impl of `std::borrow::ToOwned` for `(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync+ 'static)`
   = note: required because it appears within the type `std::borrow::Cow<'a, (dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static)>`
   = note: only the last field of a struct may have a dynamically sized type
```

And then I'm stuck :( 🆘 

Shall we have an async version of the whole `query` module? So I can protect data with mutex and all?

EDIT: Or making `body` not a `Cow` but just a reference?

Geobert committed:
WIP: stuck

Geobert commented on oxide-auth-async/src/code_grant.rs:
```diff
@@ -111,21 +111,21 @@ pub mod access_token {
         /// authorization code request.
         async fn extend(
-            &mut self, request: &dyn Request, data: Extensions,
+            &mut self, request: &(dyn Request + Sync), data: Extensions,
```
@HeroicKatora: It seems that taking a `&mut Request` forces us to have `+ Sync` which is logic and fine by me. What I don't understand, is why suddenly the `Cow<'a, dyn QueryParameter + 'static>` in both `WrappedRequest` (in `authorization` and `access_token`) asks for `Clone` trait. And we can't add as a bound as it's not an "auto trait".

Geobert replied:
This line https://github.com/Geobert/oxide-auth/blob/dfe475570e4cceb535776dd7a3deaf85e675243a/oxide-auth-async/src/endpoint/authorization.rs#L145

has this error:
```
error[E0277]: `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)` cannot be shared between threads safely
   --> oxide-auth-async\src\endpoint\authorization.rs:145:52
    |
145 |             authorization_code(&mut self.endpoint, &WrappedRequest::new(&mut request)).await;
    |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)` cannot be shared between threads safely
    |
    = help: within `endpoint::authorization::WrappedRequest<'_, R>`, the trait `std::marker::Sync` is not implemented for `(dyn oxide_auth::endpoint::query::QueryParameter + 'static)`
    = note: required because it appears within the type `&(dyn oxide_auth::endpoint::query::QueryParameter + 'static)`
    = note: required because it appears within the type `std::borrow::Cow<'_, (dyn oxide_auth::endpoint::query::QueryParameter + 'static)>`
    = note: required because it appears within the type `endpoint::authorization::WrappedRequest<'_, R>`
    = note: required for the cast to the object type `dyn oxide_auth::code_grant::authorization::Request + std::marker::Sync`
```

Which I happily want to solve by adding `+ Sync` here https://github.com/Geobert/oxide-auth/blob/dfe475570e4cceb535776dd7a3deaf85e675243a/oxide-auth-async/src/endpoint/authorization.rs#L35

And then, suddenly, `Cow` asks for: 
```
error[E0277]: the trait bound `(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static): std::clone::Clone` is not satisfied
  --> oxide-auth-async\src\endpoint\authorization.rs:35:5
   |
35 |     query: Cow<'a, dyn QueryParameter + Sync + 'static>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static)`
   |
   = note: required because of the requirements on the impl of `std::borrow::ToOwned` for `(dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static)`
   = note: required because it appears within the type `std::borrow::Cow<'a, (dyn oxide_auth::endpoint::query::QueryParameter + std::marker::Sync + 'static)>`
   = note: only the last field of a struct may have a dynamically sized type
```

Geobert replied:
Here is why, dumbby me: https://users.rust-lang.org/t/why-adding-sync-bound-asks-for-clone/46215/12

197g replied:
I'm okay with having the async primitives `Send` but I _really_ do not understand why this should influence the sync implementation in any way, and it shouldn't either. It seems that this comes from `&dyn Request` being taken as a parameter to `Extension`. This borrows it, which implies it must be `Sync` for the containing future to be `Send`.

One possible fix would be architectural: Don't use `&dyn Request` as a parameter in the async traits, instead normalize it to a specific type of Request that lives in `oxide-auth-async`, a request type that does _not_ borrow the original request and implements `Sync` and `Send`. For example, that owns the `query` after having cloned it.

Geobert replied:
OK :)

Geobert replied:
> One possible fix would be architectural: Don't use `&dyn Request` as a parameter in the async traits, instead normalize it to a specific type of Request that lives in `oxide-auth-async`, a request type that does _not_ borrow the original request and implements `Sync` and `Send`. For example, that owns the `query` after having cloned it.

That's basically `WrappedRequest` but without `Cow`, isn't it? We already have a mirror of sync WrappedRequest in async, but with the exact same fields, I'll modify them to clone the needed parts of the Request.

197g replied:
More or less. I believe that could work :)

Geobert commented on oxide-auth-async/src/endpoint/mod.rs:
```diff
@@ -93,14 +93,14 @@ pub trait OwnerSolicitor<Request: WebRequest> {
-        oxide_auth::endpoint::OwnerSolicitor::check_consent(self, req, pre_grant)
-    }
-}
+// #[async_trait]
```
I removed these blanket impl for now because it causes a wall of errors that clutter the screen

Geobert committed:
WIP: removed uneeded super trait

Geobert committed:
Remove (?Send) from async_trait

197g commented on oxide-auth-async/src/endpoint/access_token.rs:
```diff
@@ -39,12 +39,14 @@ struct WrappedToken<E: Endpoint<R>, R: WebRequest> {
-    request: PhantomData<R>,
-
+#[derive(Clone)]
+pub struct WrappedRequest<R: WebRequest>
```
The type parameter here complicates so many things.. Is it necessary? Consider that the reason it exists here is to have access to the `Error` associated type below but in that case we will never reach the `Extension` call, right?. I think all of `WrappedRequest` that is independent of the underlying request type should be in one separate struct without such a parameter.

197g commented on oxide-auth-async/src/endpoint/access_token.rs:
```diff
@@ -58,17 +60,20 @@ struct WrappedRequest<'a, R: WebRequest + 'a> {
 where
     E: Endpoint<R>,
-    R: WebRequest,
+    R: WebRequest + Send + Clone,
```
In any case, this `Clone` is never necessary. It may be necessary to implement `Clone` by-hand instead of deriving it to avoid that trait bound on the Clone impl of `WrappedRequest`.

Geobert commented:
@HeroicKatora: I manage to have something that compile, I'll try to use it before saying it's a success ;) 

It's already pushed though if you want to have a look.

197g commented on oxide-auth-async/src/endpoint/authorization.rs:
```diff
@@ -27,12 +27,14 @@ struct WrappedAuthorization<E: Endpoint<R>, R: WebRequest> {
-    request: PhantomData<R>,
-
+#[derive(Clone)]
+pub struct WrappedRequest<R: WebRequest>
```
Same here of course.

Geobert committed:
clippy fix

Geobert committed:
More Send bounds

Geobert commented:
@HeroicKatora :More Send to make compilation of my app progress.

But I'm hitting another wall though: 
```
error[E0277]: `dyn oxide_auth_async::code_grant::authorization::Endpoint<oxide_auth::frontends::simple::request::Request>` cannot be shared between threads safely
   --> src\server.rs:115:68
    |
115 |     let get_authorize = warp::path("authorize").and(warp::query()).and_then(
    |                                                                    ^^^^^^^^ `dyn oxide_auth_async::code_grant::authorization::Endpoint<oxide_auth::frontends::simple::request::Request>` cannot be shared between threads safely
    |
    = help: the trait `std::marker::Sync` is not implemented for `dyn oxide_auth_async::code_grant::authorization::Endpoint<oxide_auth::frontends::simple::request::Request>`
    = note: required because of the requirements on the impl of `std::marker::Send` for `&dyn oxide_auth_async::code_grant::authorization::Endpoint<oxide_auth::frontends::simple::request::Request>`
```

And after there is a HUGE wall of errors on Future.

But all leads to https://github.com/Geobert/oxide-auth/blob/8cdd183bf646f9ec8188fe22601e081378fb728f/oxide-auth-async/src/code_grant.rs#L395

We take `&mut dyn Endpoint<R>` which make `Sync` needed. And this time, I'm not sure that cloning will do, as inside an Endpoint there can connection to other things.

Shall we go for an Arc<Mutex<Endpoint<R>>> here?

197g commented:
`Mutex` and `async` seems like a conflict and performance killer.

197g commented on oxide-auth-async/src/code_grant.rs:
```diff
@@ -103,35 +103,52 @@ pub mod resource {
+    use crate::endpoint::access_token::WrappedRequest;
+
+    #[async_trait]
+    pub trait Extension<'a, R>
```
Not to forget that it also propagates to `Extension` which it does not in the sync code. The reasoning was that extensions to the OAuth process should be independent from the actual request type, you shouldn't need to use any data that isn't part of the usual parameter passing that's standardized.

197g reviewed (commented):
Having to add `Send` and `Sync` _to the traits_ manually still seems like a symptom of an underlying problem. Usually those get inferred. We may have to rethink the approach of taking `&dyn Endpoint` in the async flow functions as that seems to stop most of that inference. I believe if that were a concrete `impl Endpoint` instead, plus a few more similar changes, then this should work without requiring bounds everywhere. The _hard_ part is that these extra type parameters shouldn't leak into the traits. The reason is simply that this would make the traits much clumsier.

197g commented:
Writing an async library is hard :slightly_frowning_face:

Geobert commented:
> 
> 
> Writing an async library is hard 🙁

Really really hard, I wasn't expecting this level of messyness. I'm not really sure of what I'm doing, and I mostly try to follow and fix the compilation errors. As you said, there might be a deeper inherent issue here.

> I believe if that were a concrete impl Endpoint instead, plus a few more similar changes, then this should work without requiring bounds everywhere.

I don't follow the "concrete impl" part here, can you sketch an example here?

197g commented:
Currently you had this, and the type is `dyn Endpoint` which does _never_ implement `Send` on its own and `dyn Request` which is not `Sync`, so neither of `&mut dyn Endpoint` nor `&dyn Request` is `Send`.
```rust
    pub async fn access_token(
        handler: &mut dyn Endpoint, request: &dyn Request,
    ) -> Result<BearerToken, Error> {
```

Alternative one to make the future `Send` might be to make the extra bounds local to this specific future, and the flow that calls it.
```rust
    pub async fn access_token(
        handler: &mut dyn (Endpoint + Send), request: &(dyn Request + Sync),
    ) -> Result<BearerToken, Error> {
```

Alternative two turns them into type parameters so that (hopefully) they are automatically derived:
```rust
    pub async fn access_token(
        handler: &mut impl Endpoint, request: &impl Request,
    ) -> Result<BearerToken, Error> {
```

Geobert commented:
I see, I'll give that a shot next week :) Let's hope it will pan nicely

Geobert commented:
@HeroicKatora (I don't know if you get notifications if I don't mention you as this PR is on my fork, sorry if it looks "rude")

in `async::code_grant::access_token`:
```rust
#[async_trait]
    pub trait Extension {
        /// Inspect the request and extension data to produce extension data.
        ///
        /// The input data comes from the extension data produced in the handling of the
        /// authorization code request.
        async fn extend(
            &mut self, request: &impl Request, data: Extensions,
        ) -> std::result::Result<Extensions, ()>;
    }

    #[async_trait]
    impl Extension for () {
        async fn extend(
            &mut self, _: &impl Request, _: Extensions,
        ) -> std::result::Result<Extensions, ()> {
            Ok(Extensions::new())
        }
    }
```

(changed `request: WrappedRequest` for `request: &impl Request`) gives this error that I've never encounter before:
```
error[E0038]: the trait `code_grant::access_token::Extension` cannot be made into an object
   --> oxide-auth-async\src\code_grant.rs:151:36
    |
119 |     pub trait Extension {
    |               --------- this trait cannot be made into an object...
...
124 |         async fn extend(
    |                  ------ ...because method `extend` has generic type parameters       
...
151 |         fn extension(&mut self) -> &mut dyn Extension;
    |                                    ^^^^^^^^^^^^^^^^^^ the trait `code_grant::access_token::Extension` cannot be made into an object
    |
    = help: consider moving `extend` to another trait
```
So it seems that the `&impl MyTrait` route will not work or am I missing something?

EDIT: did the 1st solution (bounds on the flow and the needed func only)

Geobert committed:
New compiling state, with less bounds

Geobert commented:
I'm still trying to use it but I'm hit by https://gist.github.com/Geobert/e11a1b8918c5287896b2d4e5732583e8

Which is quite discouraging :(  I'm trying to find the cause of the first one, but I'm pulling my hairs off here :-/

Geobert committed:
Need this blanket impl for AuthMap

Geobert committed:
Some more bounds

Geobert committed:
More bounds

Geobert commented:
Is it normal/acceptable that I have to change `async::endpoint::Endpoint` and add `Sync` bounds in it like this one:
```
fn registrar(&self) -> Option<&(dyn Registrar + Sync)>;
```

197g commented:
That seems fine for the `async` endpoint specifically. It's usual in the sense that you often times you want futures to be sendable.

Geobert commented:
> 
> 
> That seems fine for the `async` endpoint specifically. It's usual in the sense that you often times you want futures to be sendable.

Thank you :) So it's the correct path to get this working, still have 15 ish wall errors. 

I'm wondering if anything is in the work on the compiler side to ease working with Future (async/await was the big step, but I feel more is needed)

Geobert committed:
Some more async Bounds

Geobert committed:
Warning fix

Geobert commented:
Can I rewrite `refresh` state machine so it doesn't borrow the request in `RefreshState`? It makes the `Sync` bounds to leak into `oxide-auth`.

197g commented:
It's the odd one out of all the statemachines right now in that regard. The style of not borrowing the request is cleaner anyways. Feel free to go ahead :)

Geobert committed:
More bounds

Geobert committed:
Modify refresh state machine to not borrow request

Geobert commented:
With these, my app compile, at last! 

Shall I fix clippy warnings as well? especially this kind of error:
```
warning: large size difference between variants
   --> C:\Users\geobe\work\oxide-auth\oxide-auth\src\code_grant\accesstoken.rs:202:5
    |
202 |     Recovered(Option<Grant>),
    |     ^^^^^^^^^^^^^^^^^^^^^^^^ this variant is 264 bytes
    |
    = note: `#[warn(clippy::large_enum_variant)]` on by default
note: and the second-largest variant is 16 bytes:
   --> C:\Users\geobe\work\oxide-auth\oxide-auth\src\code_grant\accesstoken.rs:198:5
    |
198 |     Request(&'req dyn Request),
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
help: consider boxing the large fields to reduce the total size of the enum
    |
202 |     Recovered(Box<Option<Grant>>),
    |               ^^^^^^^^^^^^^^^^^^
```

there's 3 of these

197g commented:
Looking very, very good!

> Shall I fix clippy warnings as well? especially this kind of error:

Would you mind opening a separate PR after this so that we can discuss individual ones?

Geobert commented:
> Would you mind opening a separate PR after this so that we can discuss individual ones?

I will :) 

But in order, we need to land the rebase branch, then I'll create a proper PR for sendable and then the clippy-fix PR :)

Geobert commented:
After porting some other parts (not oxide-auth dependant) to async, my oauth2 server is working again, all async! so the current state of oxide-auth-async seems correct now. Extra caution needed on `resource` flow though, as I don't use it.

Geobert committed:
never used field `grant`

Geobert committed:
fix too big size diff in variants

Geobert committed:
misc clippy fixes

Geobert committed:
fix clippy missing doc

When searching for what to say on missing doc on "request" field,
I looked at how the request was used and found that only scope() was
used, so I replace `request` with the actual used data.

Geobert committed:
fix explicit lifetime clippy

Geobert committed:
Option of Box instead of Box of Option

Commit:
Merge pull request #2 from Geobert/async-warning-cleanup

Async warning cleanup

197g committed:
Test the Send'ability of futures

Commit:
Merge pull request #3 from HeroicKatora/async-sendable

Test the Send'ability of futures

Geobert committed:
Fix resource Future being Send

Geobert committed:
restore async impl of OwnerSolicitor
